src/codegen/delete/delete_translator.cpp

```diff
@@ -0,0 +1,198 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// transaction_manager.h
+//
+// Identification: src/include/concurrency/transaction_manager.h
+//
+// Copyright (c) 2015-16, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include "codegen/delete/delete_translator.h"
+#include "common/item_pointer.h"
+#include "common/logger.h"
+#include "concurrency/transaction_manager_factory.h"
+#include "codegen/catalog_proxy.h"
+#include "codegen/data_table_proxy.h"
+#include "codegen/transaction_proxy.h"
+
+namespace peloton {
+namespace codegen {
+
+DeleteTranslator::DeleteTranslator(const planner::DeletePlan &delete_plan,
+                                   CompilationContext &context,
+                                   Pipeline &pipeline)
+    : OperatorTranslator(context, pipeline),
+      table_ptr(nullptr),
+      delete_plan_(delete_plan),
+      table_(*delete_plan_.GetTable()) {
+  // Also create the translator for our child.
+  context.Prepare(*delete_plan.GetChild(0), pipeline);
+}
+
+void DeleteTranslator::Produce() const {
+  auto &compilation_context = this->GetCompilationContext();
+
+  // the child of delete executor will be a scan. call produce function
+  // of the child to produce the scanning result
+  compilation_context.Produce(*delete_plan_.GetChild(0));
+}
+
+/**
+ * @brief The callback function that gets called for every row batch.
+ *
+ * This simply treats one row at a time, since we delete tuples one at a time.
+ *
+ * @param context The parent consumers.
+ * @param batch A reference of the code-gen'ed "current batch" value.
+ */
+void DeleteTranslator::Consume(ConsumerContext &context,
+                               RowBatch &batch) const {
+  batch.Iterate(context.GetCodeGen(),
+                [&](RowBatch::Row &row) { this->Consume(context, row); });
+}
+
+/**
+ * @brief Generate code that deals with each tuple.
+ *
+ * @param context The parent consumers.
+ * @param row A reference of the code-gen'ed "current row" value.
+ */
+void DeleteTranslator::Consume(ConsumerContext &context,
+                               RowBatch::Row &row) const {
+  auto &compilation_context = this->GetCompilationContext();
+  auto &codegen = context.GetCodeGen();
+  auto tile_group_id = row.GetTileGroupID();
+  auto tuple_id = row.GetTID(codegen);
+  llvm::Value *txn = compilation_context.GetTransactionPtr();
+
+  if (table_ptr == nullptr) {
+    storage::DataTable *table = delete_plan_.GetTable();
+
+    table_ptr = codegen.CallFunc(
+        CatalogProxy::_GetTableWithOid::GetFunction(codegen),
+        {GetCatalogPtr(), codegen.Const32(table->GetDatabaseOid()),
+         codegen.Const32(table->GetOid())});
+  }
+
+  codegen.CallFunc(_DeleteWrapper::GetFunction(codegen),
```

Comment:
It is true that the inner loop of `DELETE` is just a giant function. However, the reasoning here is that `DELETE`'s job is to talk to MVCC & Index stuff, which requires function calls anyway (unless we try to optimize MVCC & Index with LLVM also!), so we just wrap them up into a giant function. We actually have performed a timing test: by removing this giant function call (so `DELETE` doesn't actually delete anything) we reduce the running time of `DELETE` from ~260ms to ~230ms. There must be something else that's slowing this down, which we might have to do things like profiling to figure out.